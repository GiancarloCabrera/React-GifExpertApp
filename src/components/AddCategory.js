import React, { useState } from 'react';
import PropTypes from 'prop-types';


export const AddCategory = ({ setCategories }) => {
                                //Se pone un string vacio, porque si no se pone nada
                                // queda como undefined y falla
    const [inputValue, setInputValue] = useState('');
    const handleInput = (e) => {
        setInputValue(e.target.value)
        console.log(e.target.value);
    }
    const handleSubmit = (e) => {
        //Esto se usa para prevenir que el fromulario se refresque solo
        e.preventDefault();
        console.log('Formulario Hecho!!!!')
        //trim borra todos los espacios de un string
        //Si inputValue tiene mas de dos valores
        if (inputValue.trim().length > 2) {
            //El setCategories trae por defecto los valores que fueron incluidos cuando
            //el setCategories fue creado
            setCategories(cats => [inputValue, ...cats]);//inputValue contiene el valor insertado en el input
            //Ese cats contiene todos los valores almacenados hasta el momento
            //setInputValue('');
        }
    }
  return (
    //El formulario se sube cuando se presione enter
    <form onSubmit={ handleSubmit }>
        <h1>{ inputValue }</h1>
        {/* Se llama el input, el cual su valor sera lo que contenga
            el inputValue, onChange, osea cada vez que haya un cambio en ese input, se llama la funcion
            la cual le va a setear el nuevo valor escrito al inputValue, encargado de almacenar la info
            del input */}
        <input
            type="text"
            value={ inputValue }
            onChange={ handleInput }
        />
    </form>
  )
}

//La funcion setCategories es requerida obligatoriamente

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}