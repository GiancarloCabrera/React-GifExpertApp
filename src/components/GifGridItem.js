import React from 'react'
import PropTypes from 'prop-types';

export const GifGridItem = ({ img }) => {
  console.log('IMG:' ,img);
  return (
    <div className='card animate__animated animate__bounce'>
        <img src= { img.url } alt= { img.title }/>
        <p> { img.title }</p>
    </div>
  )
}

GifGridItem.propTypes = {
  img: PropTypes.object.isRequired
}
//ENZYME REQUIERE SIEMRPE QUE LOS COMPONENTES A EVALUAR SEAN EXPORTADOS 
// POR DEFAULT
export default GifGridItem;