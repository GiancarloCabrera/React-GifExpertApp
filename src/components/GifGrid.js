import React from 'react'
//import { GifGridItem } from './GifGridItem';
import { PropTypes } from 'prop-types';
//import { getGifs } from '../helpers/GetGifs'; //Se importa de helpers la funcion
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {

    //TODO ESTE CODIGO COMENTADO FUE MOVIDO A useFetchGifs DEBIDO A QUE ES UNA BUENA PRACTICA
    //ALMACENARLO EN UN HOOK

    //Creo un estado el cual sera imagenes, este recibira la info del
    // objeto gifs la cual sera mostrada por pantalla
    //const [images, setImages] = useState([]);

    //useEffect
    //Muchas veces hay codigo que necesito que corra una sola vez, asi haya cambios en el
    // en el componente. Para eso se usa el useEffect
    //En este caso react esta haciendo la request a la api cada vez que hay un camnio en el componente, y
    // yo no quiero que sea asi, necesito que solo haga la request una vez y ya, por ende voy a meter la funcion getGifs dentro
    // del hook useEffect

    // useEffect( () => {
    //     getGifs( category ).then( imgs => { //Como la funcion retorna un objeto, yo puedo acceder a el y mandarlo al setImages
    //         setImages (imgs);
    //     });
    // }, [ category ]); //Dentro de ese array van las dependencias que quiero incluir, pongo category porque cada vez que cambie quiero que ese codigo vuelva y corra

    //Esta funcion es importada del hook del archivo useFectchGifs y lo desestructuro
    const { data, loading } = useFetchGifs( category );
                        //Asi se manda una variable de este componente a otro

    return (
        <>
            <h3 className='animate__animated animate_fadeIn'>{ category }</h3>

            {/* Llamamos loading y hacemos un operador ternario,
            si loading es true muestre el parrafo, si no, no haga nada*/}
            
            { loading && <p className='animate__animated animate_flash'>LOADING</p> }

            <div className='card-grid'>
                {   //Llamamos el array data que tiene la info que queremos de la api y lo mapeamos para que mueste todo lo que queremos
                    data.map( img => {
                        //Tenemos un nuevo componente el cual esta encargado de mostrar las imagenes con sus titulos
                        return <GifGridItem 
                            key={ img.id }
                            img = { img }/>
                    })
                }
            </div>
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}
