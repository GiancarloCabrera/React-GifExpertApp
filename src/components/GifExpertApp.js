import React, { useState } from "react";
import { AddCategory } from "./AddCategory";
import { GifGrid } from "./GifGrid";

const GitExpertApp = () => {

    //const categories = ['One Punch', 'Samurai X', 'Dragon Ball'];
    //Como el array puede cambiar, puede tener mas o menos, usamos el useState
    const [categories, setCategories]= useState(['Samurai X']);
    
    //const handleAdd = () => {

        //PARA AGREGAR ELEMENTOS AL ARRAY USANDO setCategories
        //Se usa el operador spread. mantiene lo anterior y anade lo nuevo
        //setCategories([...categories, 'FIGHTER']);
        console.log(categories)

        //OTRA MANERA USANDO --------- CALLBACK
        //El cats contiene lo ya guardado y se le adiciona el LALALAND
        //setCategories(cats => [...cats, 'LALALAND'])

        //SPREADDDDDDDDDDDDDD
        //Concierte un array a una lista
    //}
    console.log(categories)

    return (
        <div>
            <h2>GitExpertApp</h2>
            {/* COMUNICACION ENTRE COMPONENTESSSSSSSSSSSSSSSSS
                Por medio del props se puede llamar el setCategories el cual 
                va a actualizar el estado del array */}

            <AddCategory setCategories= { setCategories }/>
            {/* Ese hr es una linea */}
            <hr></hr>
    
            

            {/* Mostrar un array de forma ordenada tipo lista,
                se pone la propiedad ol que es usada para listas */}
            <ol>
                {/* Con el map. lo que hace es recorrer el array y retornar
                    cada uno de los elementos del array */}
                {   //El map recibe 2 argumentos,
                    //el elemento del array y su indice
                    //No siempre es necesario poner el indice
                    categories.map( category => { 
                        //Antes se llamaba la categoria del array para mostrarla,
                        //ahora se creo un nuevo componente el cual va a mostar la categoria
                        //del array y hara una peticion http para que muestre gifs de la categoria
                        //return <li key={ category }> { category } </li>

                        //Este es el nuevo componente
                        //Recordar que REACT siempre pide un key para los valores 
                        // para poder diferenciarlos
                        return  <GifGrid 
                            key={ category }
                            category={ category }
                        />
                    })
                }
            </ol>
        </div>
    );
}

export default GitExpertApp;