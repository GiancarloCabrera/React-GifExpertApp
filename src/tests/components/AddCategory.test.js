import { shallow } from "enzyme";
import { AddCategory } from "../../components/AddCategory";

describe('Pruebas en componente AddCategory', () => { 
                        //jest.fn() es una manera de crear una simple funcion
                        // usando jest
    const setCategories = jest.fn();//Con esto se puede saber como, cuanto, cuando y donde
                                    // fue llamada la funcion
    let app;
    //Es importante llamar el beforeEach para que cada test sea aparte y una no afecte a la otra
    beforeEach( () => {
        jest.clearAllMocks(); //Esto sirve para que cada simulacion sea terminada y limpiada cada vez que termine
        app = shallow( <AddCategory setCategories={ setCategories } /> );

    });

    test('Mostrar de manera correcta el componente', () => { 
        expect( app ).toMatchSnapshot();
    }); 

    test('Debe de cambiar la caja de texto', () => { 
        const input = app.find('input');   
        const value = 'Hola Mundo';//Este value es el texto que se meteria en mi input
        //De esta manera se simula el onChange del input de mi componente
        input.simulate('change', { target: { value }});
                                //Este objeto va a tener el valor de e, 
                                //la cual es el argumento de la funcion
                                //handleInput en mi componente
    });

    test('No debe de postear la informacion con submit', () => { 
        //Encuentra el formulario en el componente y simula el onSubmit del formulario
        app.find('form').simulate('submit', {
            preventDefault: () => { } //Funcion que se va a evaluar la cual esta siendo llamada dentro del handleSubmit
        });
        //Se espera que setCategories no sea llamado porque en el input no
        // se le esta pasando nada
        expect( setCategories ).not.toHaveBeenCalled();
    });

    test('Debe de llamar el setCategories y limpiar la caja de texto', () => { 
        const value = 'Hola Mundo';

        //Simulacion del input
        app.find( 'input').simulate('change', { target: { value } } );

        //Simulacion del submit
        app.find( 'form' ).simulate('submit', { preventDefault: () => { } } );

        //Se espera que el setCategories haya sido llamado
        expect( setCategories ).toHaveBeenCalled();
        
        //Se espera que que setCategories haya sido llamado como una funcion
        expect( setCategories ).toHaveBeenCalledWith( expect.any(Function) );

        //Se espera que el valor del input sea value
        expect(app.find('input').prop('value')).toBe(value)
    });
});