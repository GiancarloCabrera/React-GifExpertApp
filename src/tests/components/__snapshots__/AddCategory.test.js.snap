// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en componente AddCategory Mostrar de manera correcta el componente 1`] = `
<form
  onSubmit={[Function]}
>
  <h1 />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
